### PR TITLE
[LoongArch] Enable FeatureExtLSX for generic-la64 processor

### PR DESCRIPTION
**Pull request by Ami-zhang**
This commit makes the `generic` target to support FP and LSX, as discussed in #110211. Thereby, it allows 128-bit vector to be enabled by default in the loongarch64 backend.

### CODE REVIEW

**Comment by heiher** [comment]
cc @xen0n

**Review by xen0n — COMMENTED** [review]
It's nice to see vectorization being leveraged across the board, as per the recommended baseline!

Note to myself: While this clearly affects `-march=generic` for Clang, I haven't verified whether `-march=loongarch64` behavior stays the same as GCC (no LSX).

2nd note to myself: GCC needs `-march=generic` implemented too for feature parity.

File: llvm/test/CodeGen/LoongArch/target-abi-from-triple-edge-cases.ll [hunk 1]
```diff
@@ -1,3 +1,4 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Comment (SixWeining) [inline llvm/test/CodeGen/LoongArch/target-abi-from-triple-edge-cases.ll:1]: The checking prefixes for this test were written manually. I prefer keeping this style unchanged this time for easy-reviewing. Then maybe you can submit a separate `NFC` PR switching to use `update_llc_test_checks.py`.

File: llvm/test/CodeGen/LoongArch/statepoint-call-lowering-r1.ll [hunk 2]
```diff
@@ -1,13 +1,15 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Comment (SixWeining) [inline llvm/test/CodeGen/LoongArch/statepoint-call-lowering-r1.ll:1]: Same as llvm/test/CodeGen/LoongArch/target-abi-from-triple-edge-cases.ll

File: llvm/test/CodeGen/LoongArch/calling-conv-lp64s.ll [hunk 3]
```diff
@@ -6,16 +6,10 @@
...
-; CHECK-NEXT:    move $a0, $a1
-; CHECK-NEXT:    bl %plt(__fixsfdi)
-; CHECK-NEXT:    add.d $a0, $fp, $a0
-; CHECK-NEXT:    ld.d $fp, $sp, 0 # 8-byte Folded Reload
-; CHECK-NEXT:    ld.d $ra, $sp, 8 # 8-byte Folded Reload
-; CHECK-NEXT:    addi.d $sp, $sp, 16
+; CHECK-NEXT:    movgr2fr.w $fa0, $a1
+; CHECK-NEXT:    ftintrz.l.s $fa0, $fa0
+; CHECK-NEXT:    movfr2gr.d $a1, $fa0
+; CHECK-NEXT:    add.d $a0, $a0, $a1
```
[the remark this answers is not in the document]
Comment (Ami-zhang) [reply llvm/test/CodeGen/LoongArch/calling-conv-lp64s.ll:12]: Updated.
Added `--mattr=-f` to stay soft-float.

File: llvm/test/CodeGen/LoongArch/statepoint-call-lowering-r1.ll [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 ; RUN: llc --mtriple=loongarch64 --verify-machineinstrs --stop-after=prologepilog < %s | FileCheck %s
 
 ;; Check that STATEPOINT instruction has an early clobber implicit def for R1.
 
 define void @test() gc "statepoint-example" {
 entry:
   %safepoint_token = tail call token (i64, i32, ptr, i32, i32, ...) @llvm.experimental.gc.statepoint.p0(i64 0, i32 0, ptr elementtype(void ()) @return_i1, i32 0, i32 0, i32 0, i32 0) ["gc-live" ()]
-; CHECK: STATEPOINT 0, 0, 0, target-flags(loongarch-call-plt) @return_i1, 2, 0, 2, 0, 2, 0, 2, 0, 2, 0, 2, 0, csr_ilp32s_lp64s, implicit-def $r3, implicit-def dead early-clobber $r1
```
[the remark this answers is not in the document]
Comment (Ami-zhang) [reply llvm/test/CodeGen/LoongArch/statepoint-call-lowering-r1.ll:8]: The automated updating of this case with `update_llc_test_checks.py` led to this behavior.
To keep the original style for easy-reviewing, I manually updated it this time.

File: llvm/test/Transforms/LoopIdiom/LoongArch/popcnt.ll [hunk 5]
```diff
@@ -2,7 +2,7 @@
 ; RUN: opt -passes=loop-idiom -mtriple=loongarch32 -mattr=+lsx -S < %s | FileCheck %s --check-prefix=CPOP
 ; RUN: opt -passes=loop-idiom -mtriple=loongarch64 -mattr=+lsx -S < %s | FileCheck %s --check-prefix=CPOP
 ; RUN: opt -passes=loop-idiom -mtriple=loongarch32 -S < %s | FileCheck %s --check-prefix=NOCPOP
-; RUN: opt -passes=loop-idiom -mtriple=loongarch64 -S < %s | FileCheck %s --check-prefix=NOCPOP
+; RUN: opt -passes=loop-idiom -mtriple=loongarch64 -S < %s | FileCheck %s --check-prefix=CPOP
```
[the remark this answers is not in the document]
Comment (Ami-zhang) [reply llvm/test/Transforms/LoopIdiom/LoongArch/popcnt.ll:5]: Updated, adjusted the testing way.

File: llvm/test/CodeGen/LoongArch/target-abi-from-triple-edge-cases.ll (hunk 1 above)
Comment (Ami-zhang) [reply llvm/test/CodeGen/LoongArch/target-abi-from-triple-edge-cases.ll:1]: Done.

File: llvm/test/CodeGen/LoongArch/statepoint-call-lowering-r1.ll (hunk 2 above)
Comment (Ami-zhang) [reply llvm/test/CodeGen/LoongArch/statepoint-call-lowering-r1.ll:1]: Done.

**Comment by Ami-zhang** [comment]
> Note to myself: While this clearly affects `-march=generic` for Clang, I haven't verified whether `-march=loongarch64` behavior stays the same as GCC (no LSX).
> 
> 2nd note to myself: GCC needs `-march=generic` implemented too for feature parity.

Thanks for your thoughtfulness and for sharing your insights! Of course, We will also continue to take note of matters related to `-march=generic`.

**Comment by xen0n** [comment]
Sorry, I misremembered things and there's actually no `-march=generic`, even in x86 land. We currently only have `loongarch64`, `la464` and `la664 for `-march` and it's okay as-is.

File: llvm/test/CodeGen/LoongArch/soft-fp-to-int.ll [hunk 6]
```diff
@@ -45,11 +45,8 @@ define i32 @fptosi_i32_double(double %X) nounwind {
 ;
 ; LA64-LABEL: fptosi_i32_double:
 ; LA64:       # %bb.0:
-; LA64-NEXT:    addi.d $sp, $sp, -16
-; LA64-NEXT:    st.d $ra, $sp, 8 # 8-byte Folded Spill
-; LA64-NEXT:    bl %plt(__fixdfsi)
-; LA64-NEXT:    ld.d $ra, $sp, 8 # 8-byte Folded Reload
-; LA64-NEXT:    addi.d $sp, $sp, 16
+; LA64-NEXT:    ftintrz.w.d $fa0, $fa0
+; LA64-NEXT:    movfr2gr.s $a0, $fa0
```
Comment (heiher) [inline llvm/test/CodeGen/LoongArch/soft-fp-to-int.ll:49]: Keeping soft-fp here.

**Comment by heiher** [comment]
This patch enables LSX by default in Clang, whereas GCC does not. Should we align the behavior with GCC?

File: llvm/test/CodeGen/LoongArch/soft-fp-to-int.ll (hunk 6 above)
Comment (Ami-zhang) [reply llvm/test/CodeGen/LoongArch/soft-fp-to-int.ll]: What I understand you to mean is that the purpose for this case should be kept, which is to test the soft-fp to int conversion.
Updated, added `--mattr=-f` to stay soft-fp.

**Comment by SixWeining** [comment]
> This patch enables LSX by default in Clang, whereas GCC does not. Should we align the behavior with GCC?

Actually it has been enabled in #100056.

Yes, clang and gcc have different mechanisms to enable lsx by default. Clang achieves this by how its source code is written while gcc achieves it by how it is configured when being built.

AFAIK, the upcoming [soft-dev-conv](https://github.com/loongson/la-softdev-convention) requires developers must enable LSX by default when building desktop and server operating systems (but not for embedded systems). Maybe maintainer of linux distribution will turn on LSX when they configure GCC.

**Review by heiher — APPROVED** [review]

**Review by xen0n — APPROVED** [review]
> AFAIK, the upcoming [soft-dev-conv](https://github.com/loongson/la-softdev-convention) requires developers must enable LSX by default when building desktop and server operating systems (but not for embedded systems).

Upcoming? Indeed it seems so because v0.1 [says](https://github.com/loongson/la-softdev-convention/blob/v0.1/la-softdev-convention.adoc#101-desktop-and-server-operating-system-build-requirements) "should".

This matters mostly from a business perspective, more so than the technical (performance) perspective, because one of the indirect consequences of *mandating* LSX (without also mandating scalar fallbacks be provided for all code) is that a body of closed-source LSX-containing components will begin to form, forcing the ecosystem building upon it to require LSX as well. This allows for easy market segmentation by e.g. providing specific CPU models without LSX/LASX -- while being lower-priced overall, they cannot run most general-purpose distributions unmodified, and even if the devs went to the great length of assembling a sysroot themselves e.g. with buildroot, they'd be still out of luck if their business depends on a closed-source blob requiring LSX. And with certain LSX/LASX features likely patent-encumbered, this makes third-party LoongArch implementations less competitive as well.

However, the status quo is already LSX all over the place -- `memcpy`/`memset` getting auto-vectorized once it's enable-by-default -- as one can verify on Compiler Explorer on on their own machine. In this case having consistency is probably better, and we should approach the embedded/indie-core problem separately.

As I have now verified several important `-march` choices to match GCC behavior for a piece of trivial loop eligible for vectorization (simple i32 adds):

* no `-march` given: LSX
* `-march=loongarch64`: scalar
* `-march=la64v1.0`: LSX

I'm accepting the patch as well. Thanks!

**Review by SixWeining — APPROVED** [review]

**Pull request merged**